Add "Modules evaluation and caching" docs (#68)
tc39/proposal-js-module-blocks: docs/module-caching.md

   ```js
   // main.js
   import { check as c1 } from "./file.js";
   import { check as c2 } from "./file.js";
   assert(c1 === c2);

   // file.js
   export const check = {};
   ```
   ```js
   // main.js
   import { check as c1 } from "./file.js";
   const { check: c2 } = await import("./file.js");
   assert(c1 === c2);
   
   // file.js
   export const check = {};
   ```
   ```js
   // main.js
   const { check: c1 } = await import("./file.js");
   const { check: c2 } = await import("./file.js");
   assert(c1 === c2);
   
   // file.js
   export const check = {};
   ```
2. Importing a module with the same string specifier from the top-level or from any module blocks defined in the same source text module and evaluated the same Realm results in a single evaluation:
   ```js
   // main.js
   import { check as c1 } from "./file.js";
   const mod = module { export { check } from "./file.js"; };
   const { check: c2 } = await import(mod);
   assert(c1 === c2);

   // file.js
   export const check = {};
   ```
   ```js
   // main.js
   const mod1 = module { export { check } from "./file.js"; };
   const mod2 = module { export { check } from "./file.js"; };

>    const { check: c2 } = await import(mod1);

Reviewer: ```suggestion
   const { check: c1 } = await import(mod1);
```

   const { check: c2 } = await import(mod2);
   assert(c2 === c2);

   // file.js
   export const check = {};
   ```
   ```js
   import { check as c1 } from "./file.js";
   const mod = module {
      const nested = module { export { check } from "./file.js"; };
      export const { check } = await import(nested);
   };
   const { check: c2 } = await import(mod);
   assert(c1 === c2);

   // file.js
   export const check = {};
   ```
3. Importing the same module block twice _from the same Realm_, even if from different scripts or modules, results in a single evaluation:
   ```js
   globalThis.count = 0;
   const mod = module { globalThis.count++; };
   await import(mod);
   await import(mod);
   assert(globalThis.count === 1);
   ```
   ```js
   // main.js
   import { mod, check as c1 } from "./dep.js";
   const { check: c2 } = await import(mod);
   assert(c1 === c2);

   // dep.js
   export const mod = module { export const check = {}; };
   export const { check } = await import(mod);
   ```
4. A module block is evaluated in the same realm as where it is imported:
   ```js
   const mod = module { globalThis.modEvaluated = true; };
   const realm = createLegacyRealm(); // [1]

>    await realm.eval(`s => import(s)`)(mod);

Reviewer: this one is very interesting. since `mod instanceof Module`, what happen if I do `import(mod)` before or after `realm.eval()`? To be more clear, this is the question:

 * `import(mod) !== realm.eval('s => import(s)')(mod)`

Reviewer: To be clear, the inconsistency here with my mental model might be coming from the fact that you might do `new Module(moduleSource)` twice, and it produces two different modules.

Author: I wrote this document a while ago and I haven't checked if it's still correct. The proposal initially cached modules based on the realm of the import call instead of the one of the module expression.

This example is probably just outdated, I have to go through the document and review it. When in doubt, please check the proposal spec text!

   const { check: c1 } = await import(mod);
   const { check: c2 } = await import(modClone);
   assert(c1 !== c2);
   assert(globalThis.count === 2);
   ```

## Invariants enforced by the HTML specification

These invariants cannot be enforced by ECMA-262, since it doesn't define how cloning works and how string specifiers are resolved. They will be respected by the HTML integration, and the champion group suggests that hosts that have a similar modules model could follow these patterns.

7. When serializing and deserializing a module block, the "base" used to resolve string specifiers (for example, the URL of the importer module) should be kept the same:
   ```js
   // /main.js
   const worker = new Worker("./worker/worker.js");
   worker.postMessage(module {
     import dir from "./foo.js";
     assert(dir === "/");
   });

   // /foo.js
   export default "/";

   // /worker/worker.js
   addEventListener("message", msg => import(msg.data));

   // /worker/foo.js
   export default "/worker";
   ```
   It follows that when importing two copies of the same module block from the same realm, transitive dependencies should only be evaluated once:
   ```js
   globalThis.count = 0;

>    const mod = module { export { check } from "./foo.js"; };

Reviewer 2: ```suggestion
   const mod = module { export { check } from "./file.js"; };
```

# Modules evaluation and caching

> This proposal aims to enforce some guarantees about how many times modules are evaluated and in which contexts. ECMAScript already guarantees that importing the same module from the same file results in a single evaluation, but Module Blocks increase the complexity because they can be passed around.

Reviewer 3: ```suggestion
This proposal aims to enforce some new guarantees around how module evaluations are cached and in which contexts. While ECMAScript already guarantees that importing the same specifier from the same parent module results in a single evaluation, Module Blocks increases the complexity by separating evaluation module records from compilation mode records and permitting compiled module records to be passed between realms. We thus need to extend the idempotency of resolution to handle these new interactions.
```

## Invariants

> 1. Importing a module with the same string specifier twice from the same file results in a single evaluation (this is already guaranteed by ecma262):

Reviewer 3: ```suggestion
1. Importing a module with the same string specifier twice from the same file results in a single evaluation (this is already guaranteed by ECMA-262):
```

   ```js
   // main.js
   import { check as c1 } from "./file.js";
   import { check as c2 } from "./file.js";
   assert(c1 === c2);
   
   // file.js
   export const check = {};
   ```
   ```js
   // main.js
   import { check as c1 } from "./file.js";
   const { check: c2 } = await import("./file.js");
   assert(c1 === c2);
   
   // file.js
   export const check = {};
   ```
   ```js
   // main.js
   const { check: c1 } = await import("./file.js");
   const { check: c2 } = await import("./file.js");
   assert(c1 === c2);
   
   // file.js
   export const check = {};
   ```

> 2. Importing a module with the same string specifier from two different modules defined in the same file and evaluated the same Realm results in a single evaluation:

Reviewer 3: ```suggestion
2. Importing a module with the same string specifier from the top-level or from any module blocks defined in the same source text module and evaluated the same Realm results in a single evaluation:
```

Perhaps for even further clarify we should include an example of a module block nested inside of another module block?

   ```js
   // main.js
   import { check as c1 } from "./file.js";
   const mod = module { export { check } from "./file.js"; };
   const { check: c2 } = await import(mod);
   assert(c1 === c2);
   
   // file.js
   export const check = {};
   ```
   ```js
   // main.js
   const mod1 = module { export { check } from "./file.js"; };
   const mod2 = module { export { check } from "./file.js"; };
   const { check: c2 } = await import(mod1);
   const { check: c2 } = await import(mod2);
   assert(c2 === c2);
   
   // file.js
   export const check = {};
   ```

> 3. Importing the same module block twice _from the same Realm_, even if from different modules, results in a single evaluation:

Reviewer 3: ```suggestion
3. Importing the same module block twice _from the same Realm_, even if from different parent modules, results in a single evaluation:
```

Author: Now that we can have syntactically nested modules we should come up with two different words for "parent" that represent a parent in the imports linking graph and a parent in the AST.

Reviewer 4: I’m confused; there shouldn’t be any such concept as “parent”, since a module can be imported many times from many different places, including itself. A module just has 0-N importers, which may be static or dynamic. (ref: node’s CJS module.parent which is a massive misfeature for this reason)

Author: Yup you are right, the word I needed is "importer".

Reviewer 3: Sounds good - `importer` or `importing` works for me. Perhaps we can update the former `referrer` terminology here as well.

   ```js
   // main.js
   import { mod, check as c1 } from "./dep.js";
   const { check: c2 } = await import(mod);
   assert(c1 === c2);

   // dep.js
   export const mod = module { export const check = {}; };
   export const { check } = await import(mod);
   ```
4. A module block is evaluated in the same realm as where it is imported:
   ```js
   const mod = module { globalThis.modEvaluated = true; };
   const realm = createLegacyRealm(); // [1]
   await realm.eval(`s => import(s)`)(mod);
   assert(globalThis.modEvaluated === undefined);
   assert(realm.modEvaluated === true);
   ```
   This is consistent with the behavior of imports with string specifiers.
5. Importing the same module block twice _from different Realms_ results in multiple evaluations:
   ```js
   const realm = createLegacyRealm(); // [1]
   globalThis.count = 0;
   realm.count = 0;
   const mod = module { globalThis.count++; export const check = {}; };
   const { check: c1 } = await import(mod);
   const { check: c2 } = await realm.eval("s => import(s)")(mod);
   assert(globalThis.count + realm.count === 2);
   assert(c1 !== c2);
   ```
   This is consistent with the behavior of imports with string specifiers.

> 6. Importing two different module blocks results in two evaluations and two different namespace objects. Two module blocks `mod1` and `mod2` are considered different if `mod1 !== mod2`:

Reviewer 5: ```suggestion
6. Importing two different module blocks with identical content results in two evaluations and two different namespace objects. Two module blocks `mod1` and `mod2` are considered different if `mod1 !== mod2`:
```

Just to make it clearer?

   ```js
   globalThis.count = 0;
   const mod1 = module { globalThis.count++; export const check = {}; };
   const mod2 = module { globalThis.count++; export const check = {}; };
   const { check: c1 } = await import(mod1);
   const { check: c1 } = await import(mod2);
   assert(c1 !== c2);
   assert(globalThis.count === 2);
   ```
   ```js
   globalThis.count = 0;
   const mod = module { globalThis.count++; export const check = {} };
   const modClone = structuredClone(mod);
   assert(mod !== modClone);
   const { check: c1 } = await import(mod);
   const { check: c2 } = await import(modClone);
   assert(c1 !== c2);
   assert(globalThis.count === 2);
   ```
   ```js
   globalThis.count = 0;
   const mod = module {};
   const worker = new Worker(`data:text/javascript,
     addEventListener("message", msg => postMessage(msg.data))
   `);
   worker.addEventListener("message", ({ data: mod2 }) => {
     assert(mod !== mod2);
     const { check: c1 } = await import(mod);
     const { check: c2 } = await import(mod2);
     assert(c1 !== c2);
     assert(globalThis.count === 2);
   });
   worker.postMessage(mod);

>    ```

Reviewer 5: Does this example add value when we have the `structuredClone`?

Author: Nope, I can remove it. The worker example is only important in (7) because it shows that resolution should be preserved even across agents.

## Invariants enforced by the HTML specification

> These invariants cannot be enforced by ecma262, since it doesn't define how cloning works and how string specifiers are resolved. They will be respected by the HTML integration, and the champion group suggests that hosts that have a similar modules model could follow these patterns.

Reviewer 3: ```suggestion
These invariants cannot be enforced by ECMA-262, since it doesn't define how cloning works and how string specifiers are resolved. They will be respected by the HTML integration, and the champion group suggests that hosts that have a similar modules model could follow these patterns.
```

> 7. When serializing&deserializing a module block, the "referrer" used as the base to resolve string specifiers should be kept the same:

Reviewer 3: ```suggestion
7. When serializing and deserializing a module block, the "referrer" used as the base to resolve string specifiers should be kept the same:
```

Reviewer 3: I think we can specify this in ECMA-262 and without needing to specify serialization or string specifier resolution.

Basically, if we can relate the concept of a `"compiled module record"` as existing across realms, then a module block can be associated against its parent source text compiled module record both of which can be fully defined ECMA-262 constructs. The resolution idempotence is then that:

> If compiled module record `b` is a module block, which is defined within a source text module compiled module record `s`, then for any realm `r` and specifier `specifier`, `HostResolveImportedModule(specifier, b, r) === HostResolveImportedModule(specifier, s, r)` where `HostResolveImportedModule` is updated to itself return a compiled module record which is then singly instanced within the realm. This is a fully-well defined invariant definition in ECMA-262. Separately, specs like compartments can then tackle the problem of multiple instancing of a compiled module within the same realm and how to relax the constraint again.

Reviewer 6: The current thinking, between @caridy and myself, is that we will bring back you `ModuleInstance` sketch, such that each module instance has its own memoization for its `importHook`. We would introduce a *_Virtual Module Record*_ that manages the lifecycle and contains a reference to a *_Static Module Record_* which, as you describe, would be immutable and safely shared between realms in the same process. *_Source Text Static Module Record_* would initially be the only concrete implementation.

That is to say, we would not alter per-realm behavior at all and that we would enforce invariance on a per-instance basis, recovering sufficient idempotence in aggregate by induction over the import hook.

Reviewer 3: It makes sense that for arbitrary resolution hooks an instance model is the idempotence model.

But it's important to note that these are two separate resolution models - one is the host default resolution model, and one is a custom resolution model.

One nice thing about maintaining the concept of a singular compartment, is that the compartment itself forms the unit of idempotence, whereas with a full instance model that boundary is no longer clear.

Therefore different constraints might apply at the uninstrumented host level, the compartment level, or the arbitrary instance level (from most constrainted idempotence to the least).

Reviewer 3: Regardless how the above works out, I think it makes sense for the host contraints for the default resolution model, and an instrumented resolution model to be clearly different resolution constraint scenarios.

Author: > Basically, if we can relate the concept of a "compiled module record" as existing across realms, then a module block can be associated against its parent source text compiled module record both of which can be fully defined ECMA-262 constructs. The resolution idempotence is then that:

How much "global" can something be? Can something also exist across agents? For example, web workers are different agents from the main one.

Author: Actually, I think that the proposed constraint is ok as-is and doesn't need to be relaxed for compartments/loaders. `structuredClone` will keep the same referrer, but every compartment can map a (_referrer_, _specifier_) to something different. This is not visible in this proposal, because currently there is 1 compartment per Realm: the cache map will need to be moved from the Realm to the Compartment.
